File: solutions/js/10.js
/**
 * @param {string} s - The string to check
 * @returns {number} The maximum depth of the magic
 */
function maxDepth(s) {
  const stash = [];
  let maxValue = 0;
  for (const char of s) {
    if (char === "[") {
      stash.push(char);
      maxValue = Math.max(maxValue, stash.length);
    } else if (!stash.pop()) {
      return -1;
    }
  }
  return stash.length > 0 ? -1 : maxValue;
}

const depth1 = maxDepth("[]"); // -> 1
console.log(depth1);

const depth2 = maxDepth("[[]]"); // -> 2
console.log(depth2);

const depth3 = maxDepth("[][]"); // -> 1
console.log(depth3);

const depth4 = maxDepth("[[][]]"); // -> 2
console.log(depth4);

const depth5 = maxDepth("[[[]]]"); // -> 3
console.log(depth5);

const depth6 = maxDepth("[][[]][]"); // -> 2
console.log(depth6);

const depth7 = maxDepth("]["); // -> -1 (closes before opening)
console.log(depth7);

const depth8 = maxDepth("[[["); // -> -1 (missing closing brackets)
console.log(depth8);

const depth9 = maxDepth("[]]]"); // -> -1 (extra closing brackets)
console.log(depth9);

const depth10 = maxDepth("[][]["); // -> -1 (one remains unclosed)
console.log(depth10);

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 140
  ops: 2155
*/

File: solutions/python/10.py
def max_depth(s: str) -> int:
    stash = []
    max_value = 0
    for char in s:
        if char == "[":
            stash.append(char)
            max_value = max(max_value, len(stash))
        elif not len(stash):
            return -1
        else:
            stash.pop()
    return -1 if len(stash) else max_value


depth1 = max_depth("[]")  # -> 1
print(depth1)

depth2 = max_depth("[[]]")  # -> 2
print(depth2)

depth3 = max_depth("[][]")  # -> 1
print(depth3)

depth4 = max_depth("[[][]]")  # -> 2
print(depth4)

depth5 = max_depth("[[[]]]")  # -> 3
print(depth5)

depth6 = max_depth("[][[]][]")  # -> 2
print(depth6)

depth7 = max_depth("][")  # -> -1 (closes before opening)
print(depth7)

depth8 = max_depth("[[[")  # -> -1 (missing closing brackets)
print(depth8)

depth9 = max_depth("[]]]")  # -> -1 (extra closing brackets)
print(depth9)

depth10 = max_depth("[][][")  # -> -1 (one remains unclosed)
print(depth10)

"""
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 140
  ops: 2235
"""
